### ISSUE TITLE
Ignored writev result value (flush functions)

### ISSUE THREAD

**Issue by perexg** [catalinii/minisatip#326]
Sometimes, when kernel buffers are full (TCP or UDP), the MPEG-TS data are silently dropped, because there's no code to requeue the data. It's problematic mainly for WAN connections when bandwidth and latency might have some low peaks, but the overall bandwidth is sufficient.

https://tvheadend.org/issues/3997

I will probably work on this some day, but if @catalinii has a little time to fix this, I would appreciate it.

**Comment by 9000h** [comment]
so similar to that?
I assume you want something like this:
```
/*
**  Handling the return value of writev is a pain.  It should return
**  the number of iov's it fully wrote out, and update the fields
**  in all of them to contain the new startpoints.
*/
int
bigwritev(fd, vp, vpcount)
    int                 fd;
    struct iovec        *vp;
    register int        vpcount;
{
    register int        i;
    register long       left;

    /* Get the total bytecount. */
    for (left = 0, i = vpcount; --i >= 0; )
        left += vp[i].iov_len;

    while (vpcount) {
        if ((i = writev(fd, vp, vpcount)) < 0)
            return -1;
        if ((left -= i) <= 0)
            break;
        for (; i >= vp->iov_len; vp++, vpcount--)
            i -= vp->iov_len;
        vp->iov_base += i;
        vp->iov_len -= i;
    }
    return 0; 
 }
```

**Comment by perexg** [comment]
Ideally, the data processing should not be blocked, but it requires more work (add another buffer for data which are not queued to the kernel and wait for space using poll/select multiplexers).

**Comment by 9000h** [comment]
ok will see what cata is thinks about the issue

**Comment by catalinii** [comment]
Hi,

thanks for the report.

I will try to work on this time permitting, but probably the best solution is to give in socketworks the ability to check for write events (with poll) and to push the data from a circular buffer if that is the case.

**Comment by estimadarocha** [comment]
@catalinii plese give a go on this issue... 

i am always getting this kind of errors:

2016-10-08 13:36:06.243 TS: DVB-S (Hispasat)/12437.14H/Baby First: H264 @ #8080 Continuity counter error (total 7382)
2016-10-08 13:36:07.768 TS: DVB-S (Hispasat)/12437.14H/Baby First: MPEG2AUDIO @ #8082 Continuity counter error (total 6741)

**Comment by lars18th** [comment]
Hi @estimadarocha ,

Your error seems to be related to "insufficient signal"... Check your antenna first!

**Comment by estimadarocha** [comment]
@lars18th no they are not... i have 3 satips the initial idea is to put each one in a friends house and them receive them all in a vps on tvheadend. When i first got these errors the first thing i do is change all the cabes.... realign lnbs and even changed them to unicable ones. So i am pretty sure thats not a signal i have in all of them more than 80/80 in signal and quality. I made some tests with the help of @perexg and the conclusion is that the problem was not related neither with tvheadend or the hardware it self. We tweaked a lot network buffers on satips and the erros decreased a lot indead its not enought yet.

To confirm all this if i put all the 3 satips runing perexg alternative firmware that uses minisatip in my own home sending to tvheadend also @home i dont get any errors.

All this tests where made in real 1000/1000 mbits vps receiving from home connections with 200/200.

**Comment by perexg** [comment]
Just a little note: @estimadarocha is using RTP/AVP/TCP transfer mode (thus no UDP loss). The config works on LAN but not on WAN. The problem is really the situation when kernel is not able to queue new data to TCP socket (the handshake through WAN needs more time than on LAN - the overall bandwidth is ok), then minisatip silently drops the written data to the socket without any error.

**Comment by lars18th** [comment]
Hi @estimadarocha & @perexg ,

OK, I know now your point. Then the problem is the WAN connection using TCP. I entirely advise against doing that! In my experience using TCP for a continous streaming is all but a pain in the ass. The best solution is a RTP over UDP... using some kind of recovering tool for lost packets. But this implies another full thread to discuss (if you like I have several proposals).

In the meantime, I suggest to comment here actions to improve the efficiency and robustness. For example:

> We tweaked a lot network buffers on satips and the erros decreased a lot indead its not enought yet.

Can you share with us these changes?

I feel is time to document all tweaks to improve the robustness of the network transmission... with that we could alleviate problems when a high resources are in use (for example with VPS, low power devices like Raspi, or with a large number of users using the same server).

What you guys think?

**Comment by estimadarocha** [comment]
i can put the vps again to work so we can make exaustive tests

the values i changed are those:

sysctl -w net.core.rmem_max=33554432
sysctl -w net.core.wmem_max=33554432
sysctl -w "net.ipv4.tcp_rmem=4096 87380 33554432"
sysctl -w "net.ipv4.tcp_wmem=4096 87380 33554432"
sysctl -w net.ipv4.tcp_window_scaling=1
sysctl -w net.ipv4.tcp_timestamps=1
sysctl -w net.ipv4.tcp_sack=1

dunno if they can be someway improved

**Comment by estimadarocha** [comment]
@perexg just a small apart do you think thats a common problem to all satip servers? even from other manufacturer

**Comment by lars18th** [comment]
Hi @estimadarocha ,

This is my custom config:

```
$  cat /etc/sysctl.conf

... (default values)

fs.file-max = 2097152
net.core.rmem_default = 8388608
net.core.rmem_max = 16777216
net.core.wmem_default = 8388608
net.core.wmem_max = 16777216
```

I don't know if your increased values of w/rmem_max has a real impact (32KB instead 16KB). I set as default 8KB.

Regarding the TCP tweaks, for sure they only have impact in the RTP/AVP/TCP transmission. In any case, I suggest to use it (TCP streaming) only in a LAN environment. Try to switch to plain RTP (UDP) transport. For use it over routed networks you only need to forward two UDP ports from your router with the publick IP to the destination machine. This is the reason for my Enhancement Request #323 . To improve the robustness I suggest to put more QoS priority in packets, like you can do with DSCP in TVHeadEnd: http://tvheadend.org/issues/2701 (while minisatip don't have this option #93 , you can use the firewall in your router to change the DSCP stream from your SAT>IP server).

Regards!

**Comment by estimadarocha** [comment]
i am not sure but i think i allready tested with udp and that also doesn't work pretty well with tvheadend

**Comment by catalinii** [comment]
Hi estimadarocha,

can you try to play with -c in minisatip. Specify -c as the bandwidth you have in KB/s. This should make sure you are not pushing more than needed per second. I had a similar issue in the past where the router of the network where minisatip was running was slow and running on pppoe connection.

Thanks

**Comment by lars18th** [comment]
Hi @estimadarocha ,

> i am not sure but i think i allready tested with udp and that also doesn't work pretty well with tvheadend

Using a different router can help you. Consumer grade routers (almost) doesn't have queues; however some of them have a kind of QoS. In any case I recommend to test with Mikrotik routers. You can tweak your RTP/UDP connection for good transmit. I do several checks and it only lacks robustness over missing packets (over Internet it's impossible to grant end-to-end QoS, so you need to protect over missing packets).

I hope this helps.

**Comment by perexg** [comment]
Well guys, all these suggestions have more impact than the fix for TCP queuing in minisatip (and looking to the code - UDP is affected, too). It's certainly only software issue - the kernel will retransmit packets for TCP. RTP/UDP is unreliable by design. It works mostly OK for low-bandwidth data like telephony, but it's not good for DVB streams. And you require perfect per-packet (not block) queuing to spread outgoing packets linearly in the time - it's difficult without the RT kernel. Even on LAN you may hit packet loss when your switch / ethernet card decides to drop one ethernet packet.

**Comment by estimadarocha** [comment]
@catalinii i did and no luck on that... those problems happen even with a single channel streaming so guess its not a bandwitch limitation issue.

@lars18th i can't change my main router thats the isp fiber router... but in one of my friends house he was a non consumer grade one and problem happens with both tcp/udp.

@perexg a permanent solution will be very very welcome here, if this happen will you integrate it on satipaxe?

Thanks to all i really apreciate the help in this

**Comment by lars18th** [comment]
Hi @perexg ,

> RTP/UDP is unreliable by design. It works mostly OK for low-bandwidth data like telephony, 
> but it's not good for DVB streams. 

Yes, RTP/UDP is unreliable. But I don't agree that it's not good for MPEG-TS... Why then are used in professional equipment? Why then the SAT>IP protocol works "perfect" on LAN wired networks? The problem isn't only the protocol, other factors have also influence, for example the free bandwith.

> Even on LAN you may hit packet loss when your switch / ethernet 
> card decides to drop one ethernet packet.

Yes, this is the real problem of MPEG-TS over IP instad of over RF. The packet loss has terrible effects, more than a bit/burst errors. But IP networks only deliver the full packet or nothing (please, don't add UDPlite to the equation).

So, I suggest to improve the network efficiency of the software, but also improve the network equipment. And regarding the WAN transmission... other techniques can be applied (FEC or Packet Recovery).

**Comment by lars18th** [comment]
Hi @estimadarocha ,

> @lars18th i can't change my main router thats the isp fiber router...

You can use a second router in half-bridge mode... the main router then can disable the firewall and in fact it will be transparent. Or you can use it as a bridge, if your ISP uses PPPoE and then use it only as a modem. Or if you have one ONT and one router, then you can replace the router.
 
> but in one of my friends house he was a non consumer grade one 
> and problem happens with both tcp/udp.

Then the problem is the ISP: it discards several packets over the WAN connection. You can try to use HLS for inter-house connection... but not for realtime streaming, only for recordings.

**Comment by estimadarocha** [comment]
this are the kind of errors in tvheadend this night:

2016-10-14 22:11:53.584 TS: DVB-S (Hispasat)/12206.98H/SPORT TV 1: H264 @ #5129: Invalid start code f3:03:e5
2016-10-14 22:12:03.809 TS: DVB-S (Hispasat)/12206.98H/SPORT TV 1: H264 @ #5129: Invalid start code 30:f1:8a
2016-10-14 22:12:06.325 descrambler: ECM - key late (1250 ms) for service "SPORT TV 1"

as you can see descrambler even fail because of mess with ecm's!!!

2016-10-14 22:14:50.179 TS: DVB-S (Hispasat)/12206.98H/SPORT TV 1: H264 @ #5129: Invalid start code e8:e4:bc
2016-10-14 22:15:00.183 TS: DVB-S (Hispasat)/12206.98H/SPORT TV 1: H264 @ #5129: Invalid start code db:15:5b
2016-10-14 22:15:00.183 TS: DVB-S (Hispasat)/12206.98H/SPORT TV 1: AAC @ #5130 Corrupted PES header (errors 1)
2016-10-14 22:15:10.237 TS: DVB-S (Hispasat)/12206.98H/SPORT TV 1: H264 @ #5129: Invalid start code eb:e8:10

and a bunch of these

**Comment by estimadarocha** [comment]
and also this:

2016-10-14 20:55:26.556 TS: DVB-S (Hispasat)/12245.34H/Sport TV 1 HD: H264 @ #6433: Invalid start code ce:28:c1
2016-10-14 20:55:26.559 TS: DVB-S (Hispasat)/12245.34H/Sport TV 1 HD: MPEG2AUDIO @ #6434: Invalid start code 28:59:64
2016-10-14 20:55:28.220 subscription: 00E7: service instance is bad, reason: No input detected

**Comment by estimadarocha** [comment]
@catalinii any new ideas for this?

**Comment by catalinii** [comment]
Hi,

I will try to implement this mid next month...and will let you know when I have a version ready.

Thanks

**Comment by estimadarocha** [comment]
@catalinii thanks

your help is really apreciated... this weekend i even got these errors working just on internal network

**Comment by estimadarocha** [comment]
Thanks i really apreciate the effort

Thanks

Em 23/10/2016 15:53, "catalinii" <notifications@github.com> escreveu:

**Comment by catalinii** [comment]
Hi,

test and let me know how it works now.

You can increase the buffer with -B.

Check for messages like:

closing stream 0
sockets_del: 11 -> handle 11, sid 0, **overflow** 0

If it is > 0  increase the buffer.

Thanks

**Comment by estimadarocha** [comment]
Hi @catalinii... you are the man... now i have to wait @perexg to compile satipaxe with it!

Was it easy to handle the problem? 

By the way when you said i contact you by email u mean minisatip@noreply.github.com???

**Comment by catalinii** [comment]
Hi,

As you can see the commit is consistent... but still not fully tested atm.
I meant the email from the source code: https://github.com/catalinii/minisatip/blob/master/adapter.c

Thanks

**Comment by catalinii** [comment]
this is fixed in the 0.7.4.. closing
